doc/neps/nep-0029-dtype-as-type.rst

```diff
@@ -0,0 +1,268 @@
+===============================================
+NEP 29 — Refactor Dtypes to become Type Objects
+===============================================
+
+:Author: Matti Picus
+:Status: Draft
+:Type: Standards Track
+:Created: 2018-12-27
+
+
+Abstract
+========
+
+NumPy's `dtype <http://www.numpy.org/devdocs/reference/generated/numpy.dtype.html>`
+is a python class with the simple ``mro`` ``[np.dtype, object]``. Creating an
+instance of ``dtype`` *i.e.* ``a = np.dtype('int8')`` will result in a python
+object of type ``dtype``. The ``dtype`` obect instance has attributes, among
+them ``a.type``, which is a class object. Instantiating that class object
+``a.type(3)`` produces a numpy `scalar
+<http://www.numpy.org/devdocs/reference/arrays.scalars.html>`.
+
+This NEP proposes a different class heirarchy. Objects of ``np.dtype`` will
+become type objects with a heirarchical ``mro`` like scalars. They will support
+subclassing. A future NEP may propose that instantiating a dtype type object
+will produce a scalar refleting that dtype, but that is not a goal of this NEP.
+
+The changed dtype will:
+
+- facilitate extending dtypes, typically for things like categoricals, novel
+  representations like datetime or IP addresses, or adding attributes like
+  units.
+- Simplify the code around ``__repr__`` and method lookup.
+
+Overall Design
+--------------
+
+In pure python (without error checking)::
+
+    import numpy as np
+
+    class Dtype(type):
+
+        def __new__(cls, obj, *args, **kwargs):
+            if isinstance(obj, int):
+                return dtype_int_dict[obj]
+            elif isinstance(obj, type) and issubclass(obj, np.generic):
+                return dtype_scalar_dict[obj]
+            elif len(args) < 1:
+                # Dtype('int8') or Dtype('S10') or record descr
+                return create_new_descr(cls, obj, *args, **kwargs)
+            else:
+                return super().__new__(cls, obj, *args, **kwargs)
+
+        def __call__(self, *args, **kwargs):
+            return self.typeobj(*args, **kwargs)
+
+    class IntDtype(Dtype):
+        def __repr__(self):
+            if self is IntDescr:
+                return type.__repr__(self)
+            return 'dtype(%s%d)' %(self.kind, self.itemsize)
+
+    class GenericDescr(type, metaclass=Dtype):
+        pass
+
+    class IntDescr(GenericDescr, metaclass=IntDtype):
+        def format(value):
```

Comment:
For future proofing, I think this might as well be
```
def __format__(self, format_spec=''):
    return format(int(self), format_spec)
```

Reply:
Actually, following that logic, logically perhaps for now this should be
```
def __str__(self):
    return str(int(self))
```

Reply:
In this design, `format` is an unbound class function, it would be called without a self to format the value of the memory location when printing an `ndarray`

Reply:
ok